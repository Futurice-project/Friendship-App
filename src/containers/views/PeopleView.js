import React from 'react';
import { connect } from 'react-redux';
import { View, FlatList, ActivityIndicator } from 'react-native';
import { NavigationActions } from 'react-navigation';
import rest from '../../utils/rest';
import { SearchBar } from 'react-native-elements';
import throttle from 'lodash/throttle';

import { Title, Header, SmallHeader, Description } from '../../components/Text';

import { ViewContainerTop, ViewContainer, Centered, FullscreenCentered, IconImage } from '../../components/Layout';
import Person from '../../components/Person';
import Tag from '../../components/Tags';
import RoundTab from '../../components/RoundTab';
import Spinner from '../../components/Spinner';

const mapStateToProps = state => ({
  usersSearch: state.usersSearch,
});

const mapDispatchToProps = dispatch => ({
  refreshUsersSearch: username => {
    /* .force() abort previous request if it performs and after that perform new request. This
    method combines abort and direct call action methods. it prevent a warning about unhandled
    promises rejection */
    dispatch(rest.actions.usersSearch.force({ username }));
  },
});

export class PeopleView extends React.Component {
  static navigationOptions = {
    tabBarIcon: ({ tintColor }) => (
      <IconImage
        source={require('../../../assets/search0.png')}
        tintColor={tintColor}
      />
    ),
  };

  state = {
    data: [],

    searchedUsername: '',
    currentPage: 0,
    loading: false,
    filteredUsers: [],
    searchedUsername: '',
    infiniteScrollStop: false,
  };

  componentDidMount() {
    this.fetchData();
  }

  // fetch 10 users and add them to the state.data
  fetchData = () => {
    this.setState({ loading: true });
    fetch('http://localhost:3888/users/page/' + this.state.currentPage)
      .then(response => {
        return response.json();
      })
      .then(response => {
        this.setState({ currentPage: this.state.currentPage + 1 });
        this.setState({ data: [...this.state.data, ...response], loading: false});
      })
      .catch(err => console.error(err + ' error fetchData in peopleView.js'));
  };

  handleEnd = () => {
    if (!this.onEndReachedCalledDuringMomentum) {
      // fetch 10 more users from the db
      this.fetchData();
      this.onEndReachedCalledDuringMomentum = true;
    }
  };

  // Creates a throttled function that only invokes func at most once per every 1 second.
  getUserByUsername = throttle(username => {
    this.setState({ searchedUsername: username, infiniteScrollStop: username ? true : false, });
    this.props.refreshUsersSearch(username);
  }, 1000);

  renderPeople() {
    if (this.props.usersSearch.loading) {
      return <ActivityIndicator />;
    }
    return (
      <View>
        <RoundTab tint="#ffffff" title="PEOPLE" />
        <Centered style={{ paddingBottom: 45, backgroundColor: '#fff' }}>
          <FlatList
            data={
              this.state.searchedUsername.length > 0 ? (
                this.props.usersSearch.data
              ) : (
                this.state.data
              )
            }
            keyExtractor={item => item.id}
            renderItem={({ item }) => <Person box data={item} />}
            onEndReached={this.handleEnd}
            onEndReachedThreshold={0.4}
            onMomentumScrollBegin={() => {
              this.onEndReachedCalledDuringMomentum = false;
            }}
            horizontal
          />
        </Centered>
      </View>
    );
  }
  
  render() {
    return (
      <ViewContainerTop style={{ backgroundColor: '#e8e9e8' }}>
        <SearchBar
          lightTheme
          containerStyle={{
            backgroundColor: '#e8e9e8',
            borderTopColor: '#e8e9e8',
            borderBottomColor: '#e8e9e8',
            marginVertical: 10,
            marginHorizontal: 5,
          }}
          inputStyle={{ backgroundColor: '#fff' }}
          onChangeText={username => this.getUserByUsername(username)}
          autoCapitalize="none"
          autoCorrect={false}
          placeholder="Search People"
          clearIcon
        />
        {this.renderPeople()}
      </ViewContainerTop>
    );
  }

  render = () => (
    <ViewContainer>
      <Title> People </Title>
      <SearchBar
        round
        lightTheme
        onChangeText={username => this.getUserByUsername(username)}
        placeholder="Search"
      />

      <FullscreenCentered>
        <FlatList
          data={
            this.state.searchedUsername.length > 0
              ? this.state.filteredUsers
              : this.state.data
          }
          keyExtractor={this.keyExtractor}
          renderItem={this.renderItem}
          onEndReached={this.handleEnd}
          onEndReachedThreshold={0.4}
          style={{ flex: 1 }}
          //ListFooterComponent= {() => <ActivityIndicator animating size= 'small'/>}
          horizontal
        />
        {this.renderSpinner()}
      </FullscreenCentered>
    </ViewContainer>
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(PeopleView);
