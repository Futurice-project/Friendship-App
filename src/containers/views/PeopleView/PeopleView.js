import React from 'react';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import rest from '../../../utils/rest';
import throttle from 'lodash/throttle';
import Background from '../../../components/Background/Background';
import PeopleList from '../../../components/PeopleList';

const mapStateToProps = state => ({
  usersSearch: state.usersSearch,
  auth: state.auth,
});

const mapDispatchToProps = dispatch => ({
  refreshUsersSearch: username => {
    /* .force() abort previous request if it performs and after that perform new request. This
    method combines abort and direct call action methods. it prevent a warning about unhandled
    promises rejection */
    dispatch(rest.actions.usersSearch.force({ username }));
  },
  redirectToWelcomeScreen: () =>
    dispatch(
      NavigationActions.reset({
        index: 0,
        actions: [NavigationActions.navigate({ routeName: 'Welcome' })],
      }),
    ),
});

export class PeopleView extends React.Component {
  state = {
    searchedUsername: '',
  };

  componentDidMount() {
    this.redirectWhenNotLoggedIn();
    /*this.fetchUsersForPage(this.state.currentPage);*/
  }

  componentWillReceiveProps(nextProps) {
    this.setStateWithUsersData(nextProps);
  }

  redirectWhenNotLoggedIn = () => {
    if (!this.props.auth.data.decoded) {
      this.props.redirectToWelcomeScreen();
    }
  };

  setStateWithUsersData = nextProps => {
    if (
      nextProps.usersByPage &&
      nextProps.usersByPage.data &&
      nextProps.usersByPage.data !== this.props.usersByPage.data
    ) {
      this.setState({
        userData: [...this.state.userData, ...nextProps.usersByPage.data],
      });
    }
  };

  // Creates a throttled function that only invokes refreshUsersSearch at most once per every 1 second.
  getUserByUsername = throttle(username => {
    this.setState({ searchedUsername: username });
    this.props.refreshUsersSearch(username);
  }, 1000);

  render() {
    return (
      <Background color="grey">
        <PeopleList />
      </Background>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(PeopleView);
