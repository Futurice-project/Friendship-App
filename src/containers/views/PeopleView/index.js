import PeopleView from './PeopleView';

export default PeopleView;
