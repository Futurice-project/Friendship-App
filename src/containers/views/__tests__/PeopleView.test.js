import React from 'react';
import 'react-native';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import PeopleView from '../PeopleView';

describe('PeopleView', () => {
  const mockStore = configureMockStore([thunk]);

  const users = [
    {
      id: 2,
      email: 'test@test.com',
      username: 'justATest',
      avatar: 'https://friendship-app.s3.amazonaws.com/avatars/avatar1.png',
      birthyear: 1980,
      genderlist: ['HUMAN'],
      loveCommon: 2,
      hateCommon: 1,
      locations: ['Istanbul'],
    },
    {
      id: 3,
      email: 'test@test.com',
      username: 'justATest',
      avatar: 'https://friendship-app.s3.amazonaws.com/avatars/avatar1.png',
      birthyear: 1980,
      genderlist: ['HUMAN'],
      loveCommon: 2,
      hateCommon: 1,
      locations: ['Istanbul'],
    },
  ];

  const initialState = {
    usersByPage: {
      data: {
        data: users,
      },
    },
    auth: {
      data: {
        decoded: {
          id: 1,
        },
      },
    },
  };

  test('renders correctly', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(<PeopleView store={store} />);

    expect(wrapper).toMatchSnapshot();
  });
});
