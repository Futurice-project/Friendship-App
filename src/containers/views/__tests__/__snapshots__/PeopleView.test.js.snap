// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeopleView renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(PeopleView)
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "auth": Object {
        "data": Object {
          "decoded": Object {
            "id": 1,
          },
        },
      },
      "fetchUsersByPage": [Function],
      "redirectToWelcomeScreen": [Function],
      "refreshUsersSearch": [Function],
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
      "usersByPage": Object {
        "data": Object {
          "data": Array [
            Object {
              "birthyear": 1980,
              "email": "test@test.com",
              "emoji": "ï¿½ï¿½ðŸ‡­",
              "genderlist": Array [
                "HUMAN",
              ],
              "hateCommon": 1,
              "id": 2,
              "locations": Array [
                "Istanbul",
              ],
              "loveCommon": 2,
              "username": "justATest",
            },
            Object {
              "birthyear": 1980,
              "email": "test@test.com",
              "emoji": "ï¿½ï¿½ðŸ‡­",
              "genderlist": Array [
                "HUMAN",
              ],
              "hateCommon": 1,
              "id": 3,
              "locations": Array [
                "Istanbul",
              ],
              "loveCommon": 2,
              "username": "justATest",
            },
          ],
        },
      },
      "usersSearch": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "auth": Object {
          "data": Object {
            "decoded": Object {
              "id": 1,
            },
          },
        },
        "fetchUsersByPage": [Function],
        "redirectToWelcomeScreen": [Function],
        "refreshUsersSearch": [Function],
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
        "usersByPage": Object {
          "data": Object {
            "data": Array [
              Object {
                "birthyear": 1980,
                "email": "test@test.com",
                "emoji": "ï¿½ï¿½ðŸ‡­",
                "genderlist": Array [
                  "HUMAN",
                ],
                "hateCommon": 1,
                "id": 2,
                "locations": Array [
                  "Istanbul",
                ],
                "loveCommon": 2,
                "username": "justATest",
              },
              Object {
                "birthyear": 1980,
                "email": "test@test.com",
                "emoji": "ï¿½ï¿½ðŸ‡­",
                "genderlist": Array [
                  "HUMAN",
                ],
                "hateCommon": 1,
                "id": 3,
                "locations": Array [
                  "Istanbul",
                ],
                "loveCommon": 2,
                "username": "justATest",
              },
            ],
          },
        },
        "usersSearch": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
